### PR TITLE
Add sum-zero constrained intrinsic conditional autoregressive (ICAR) variable

### PR DESCRIPTION
**Pull request by zoj613**
THis PR is a followup on https://github.com/pymc-devs/pymc3/issues/4518#issuecomment-811478578 . It impelements the sum-zero constrained intrinsic conditonal autoregressive (ICAR) model as described in [1].

References
----------
[1] Keefe M.J., Ferreira M.A.R., Franck C.T. On the formal specification of sum-zero constrained intrinsic conditional autoregressive models Spat. Stat., 24 (2018), pp. 54-65

Depending on what your PR does, here are a few things you might want to address in the description:
+ [x] important background, or details about the implementation
+ [x] are the changes—especially new features—covered by tests and docstrings?
+ [x] [linting/style checks have been run](https://github.com/pymc-devs/pymc3/wiki/PyMC3-Python-Code-Style)
+ [ ] [consider adding/updating relevant example notebooks](https://github.com/pymc-devs/pymc-examples)
+ [ ] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

### CODE REVIEW

**Comment by zoj613** [comment]
Copy of the article can be found here: https://github.com/pymc-devs/pymc3/issues/4518#issuecomment-811532299

File: pymc3/distributions/multivariate.py [hunk 1]
```diff
@@ -2048,3 +2049,51 @@ def random(self, point=None, size=None):
...
+    """
+    rv_op = icar
+
+    @classmethod
+    def dist(cls, A, tau, *args, **kwargs):
+        return super().dist([A, tau], **kwargs)
+
+    def logp(value, A, tau):
+        A = at.as_tensor_variable(A, ndim=2)
+        tau = at.as_tensor_variable(tau, ndim=0)
```
Comment (ricardoV94) [inline pymc3/distributions/multivariate.py]:
```suggestion
    def dist(cls, A, tau, *args, **kwargs):
        A = at.as_tensor_variable(floatX(A), ndim=2)
        tau = at.as_tensor_variable(floatX(tau), ndim=0)
        return super().dist([A, tau], **kwargs)

    def logp(value, A, tau):
```

File: pymc3/distributions/multivariate.py [hunk 2]
```diff
@@ -2048,3 +2049,51 @@ def random(self, point=None, size=None):
...
+        d = eigvalsh(W, at.eye(W.shape[0]))[1:]
+        out = (
+            0.5 * (n - 1) * at.log(2 * np.pi * tau)
+            + 0.5 * at.log(d).sum()
+            - 0.5 * tau * at.dot(value, at.dot(W, value))
+        )
+        return bound(out, at.math.abs_(at.sum(value)) < 1e-06)
+
+    def logcdf(value, tau, cov):
+        raise NotImplementedError()
```
Comment (ricardoV94) [inline pymc3/distributions/multivariate.py]:
```suggestion
```

That's not needed

File: pymc3/distributions/multivariate.py [hunk 3]
```diff
@@ -2048,3 +2049,51 @@ def random(self, point=None, size=None):
...
+        cov = np.linalg.pinv(W) / tau
+        return multivariate_normal.rng_fn(rng=rng, mean=mu, cov=cov, size=size)
+
+
+icar = ICARRV()
+
+
+class ICAR(Continuous):
+    r"""
+    The zero-sum constrained Intrinsic Conditional Autoregressive Model.
```
Comment (ricardoV94) [inline pymc3/distributions/multivariate.py:2127]: These docstrings need to be expanded. Also the distribution has to be included in the docs.

**Review by ricardoV94 — COMMENTED** [review]
Thanks for opening the PR. I left some suggestions.

I suggest you check this developer guide for the next steps (docstrings, tests, etc...): https://www.github.com/pymc-devs/pymc3/tree/main/docs%2Fsource%2Fdeveloper_guide_implementing_distribution.md

File: pymc3/distributions/multivariate.py (hunk 3 above)
Comment (zoj613) [reply pymc3/distributions/multivariate.py:2127]: I plan to expand on it as soon as I get everything working.

**Comment by zoj613** [comment]
@ricardoV94  is there infrastructure to generate adjecency matrices for testing CAR variables? I'm having trouble creating input for the `test_distributions.py` test im supposed to write.

Normally, i'd use something like the following, assuming the area of interest is made up of a rectangular lattice:
```python
import libpysal 

lattice_rows = 2
lattice_cols = 5
A = libpysal.weights.lat2SW(lattice_rows, lattice_cols, criterion='rook', row_st=False)
```

which gives:
```python
>>> A.toarray()
array([[0, 1, 0, 0, 0, 1, 0, 0, 0, 0],
       [1, 0, 1, 0, 0, 0, 1, 0, 0, 0],
       [0, 1, 0, 1, 0, 0, 0, 1, 0, 0],
       [0, 0, 1, 0, 1, 0, 0, 0, 1, 0],
       [0, 0, 0, 1, 0, 0, 0, 0, 0, 1],
       [1, 0, 0, 0, 0, 0, 1, 0, 0, 0],
       [0, 1, 0, 0, 0, 1, 0, 1, 0, 0],
       [0, 0, 1, 0, 0, 0, 1, 0, 1, 0],
       [0, 0, 0, 1, 0, 0, 0, 1, 0, 1],
       [0, 0, 0, 0, 1, 0, 0, 0, 1, 0]], dtype=int8)
```

**Comment by ricardoV94** [comment]
You can hardcode some predefined matrices if it's too complicated to generate then dynamically. If it's too many you can put them in a separate file.

**Comment by zoj613** [comment]
> You can hardcode some predefined matrices if it's too complicated to generate then dynamically. If it's too many you can put them in a separate file.

I added a simple function to generate such, though it's far from being robust. I am now getting weird errors locally for the `logp` tests. I think maybe my understanding of the `Domain` objects for multivariate distributions is lacking. I am getting ` IndexError: list index out of range` and 
`

<details>
<summary>Details</summary>

```
x = A

    def _is_sparse(x):
        """
    
        Returns
        -------
        boolean
            True iff x is a L{scipy.sparse.spmatrix} (and not a L{numpy.ndarray}).
    
        """
        if not isinstance(x, (scipy.sparse.spmatrix, np.ndarray, tuple, list)):
>           raise NotImplementedError(
                "this function should only be called on "
                "sparse.scipy.sparse.spmatrix or "
                "numpy.ndarray, not,",
                x,
            )
E           NotImplementedError: ('this function should only be called on sparse.scipy.sparse.spmatrix or numpy.ndarray, not,', A)
```

</details>

depending on how I define the input to `check_logp`

**Comment by zoj613** [comment]
@ricardoV94 any suggestions regarding the above errors?

**Comment by ricardoV94** [comment]
> @ricardoV94 any suggestions regarding the above errors?

The edge entries in a `Domain` are excluded. If you have less than three entries (e.g., `Domain([0, 1]`), there will be nothing usable in the `check_logp`. Could this be your issue? For instance, of your `tau` domain only the middle value will ever be tested.

**Comment by zoj613** [comment]
> > @ricardoV94 any suggestions regarding the above errors?
> 
> The edge entries in a `Domain` are excluded. If you have less than three entries (e.g., `Domain([0, 1]`), there will be nothing usable in the `check_logp`. Could this be your issue? For instance, of your `tau` domain only the middle value will ever be tested.

I tried that using the following:
```python
        adj_mat = Domain([random_adjacency_matrix(n, rng) for _ in range(5)])
        dom = Domain([generate_vals(n, rng) for _ in range(5)])
        tau = Domain([0.1, 0.25, 1., 1.5, 2.])
```

But raises the first error I got, which is: `NotImplementedError: ('this function should only be called on sparse.scipy.sparse.spmatrix or numpy.ndarray, not,', A)`. It appears that the `aesara.tensor.sharedvar.TensorSharedVariable` arguments are not passed in correctly to the `dist` function.

See the full trace below:

<details>
<summary>Details</summary>

```
...
        adj_mat = Domain([random_adjacency_matrix(n, rng) for _ in range(5)])
        dom = Domain([generate_vals(n, rng) for _ in range(5)])
        tau = Domain([0.1, 0.25, 1., 1.5, 2.])
    
>       self.check_logp(
            ICAR,
            dom,
            {"A": adj_mat, "tau": tau},
            ref_icar_logp,
        )

pymc3/tests/test_distributions.py:2088: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pymc3/tests/test_distributions.py:657: in check_logp
    model, param_vars = build_model(pymc3_dist, domain, paramdomains, extra_args)
pymc3/tests/test_distributions.py:249: in build_model
    distfam(
pymc3/distributions/distribution.py:207: in __new__
    rv_out = cls.dist(*args, rng=rng, initval=None, **kwargs)
pymc3/distributions/multivariate.py:2130: in dist
    if aesara.sparse._is_sparse(A):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

x = A

    def _is_sparse(x):
        """
    
        Returns
        -------
        boolean
            True iff x is a L{scipy.sparse.spmatrix} (and not a L{numpy.ndarray}).
    
        """
        if not isinstance(x, (scipy.sparse.spmatrix, np.ndarray, tuple, list)):
>           raise NotImplementedError(
                "this function should only be called on "
                "sparse.scipy.sparse.spmatrix or "
                "numpy.ndarray, not,",
                x,
            )
E           NotImplementedError: ('this function should only be called on sparse.scipy.sparse.spmatrix or numpy.ndarray, not,', A)

../.pyenv/versions/miniconda3-4.7.12/envs/pymc3/lib/python3.9/site-packages/aesara/sparse/type.py:18: NotImplementedError
=========================================================================================================== short test summary info ===========================================================================================================
FAILED pymc3/tests/test_distributions.py::TestMatchesScipy::test_icar[3] - NotImplementedError: ('this function should only be called on sparse.scipy.sparse.spmatrix or numpy.ndarray, not,', A)
FAILED pymc3/tests/test_distributions.py::TestMatchesScipy::test_icar[20] - NotImplementedError: ('this function should only be called on sparse.scipy.sparse.spmatrix or numpy.ndarray, not,', A)
FAILED pymc3/tests/test_distributions.py::TestMatchesScipy::test_icar[100] - NotImplementedError: ('this function should only be called on sparse.scipy.sparse.spmatrix or numpy.ndarray, not,', A)
FAILED pymc3/tests/test_distributions.py::TestMatchesScipy::test_icar[500] - NotImplementedError: ('this function should only be called on sparse.scipy.sparse.spmatrix or numpy.ndarray, not,', A)

```

</details>



EDIT: if I comment out the lines:
```
+        #if aesara.sparse._is_sparse(A):
+        #    A = aesara.sparse.dense_from_sparse(floatX(A))
+        #else:
+        #    A = at.as_tensor_variable(floatX(A))
+        A = at.as_tensor_variable(floatX(A))
```

the the tests run as expected. Is there a subtle bug in `_is_sparse` that cannot handle certain input?


EDIT2: I decided to use `_is_sparse_variable` instead and things work now.

**Comment by zoj613** [comment]
I think this is ready for a round of review @brandonwillard @fonnesbeck @ricardoV94 

The generation of the adjacency matrix can probably be improved by a lot to prevent the weird failures in tests.

File: pymc3/distributions/multivariate.py [hunk 4]
```diff
@@ -2048,3 +2050,100 @@ def random(self, point=None, size=None):
...
+    @classmethod
+    def dist(cls, A, tau, *args, **kwargs):
+        A = aesara.sparse.as_sparse_or_tensor_variable(floatX(A))
+        if aesara.sparse.basic._is_sparse_variable(A):
+            A = aesara.sparse.dense_from_sparse(floatX(A))
+        else:
+            A = at.as_tensor_variable(floatX(A))
+        tau = at.as_tensor_variable(floatX(tau))
+        Assert("`A` needs to be a 2D tensor")(at.eq(A.ndim, 2))
+        Assert("`tau` needs to be a 0D tensor")(at.eq(tau.ndim, 0))
```
Comment (brandonwillard) [inline pymc3/distributions/multivariate.py]: The output of these `Assert`s needs to be in the graph for them to take effect.

```suggestion
        A = Assert("`A` needs to be a 2D tensor")(A, at.eq(A.ndim, 2))
        tau = Assert("`tau` needs to be a 0D tensor")(tau, at.eq(tau.ndim, 0))
```

File: pymc3/distributions/multivariate.py [hunk 5]
```diff
@@ -2048,3 +2050,100 @@ def random(self, point=None, size=None):
...
+    def logp(value, A, tau):
+        n = A.shape[0]
+        W = at.diag(A.sum(axis=0)) - A
+        d = eigvalsh(W, at.eye(W.shape[0]))[1:]
+        out = (
+            0.5 * (n - 1) * at.log(2 * np.pi * tau)
+            + 0.5 * at.log(d).sum()
+            - 0.5 * tau * at.dot(value, at.dot(W, value))
+        )
+        return bound(out, at.math.abs_(at.sum(value)) < 1e-06)
```
Comment (brandonwillard) [inline pymc3/distributions/multivariate.py]: Is this effectively being clipped for numerical reasons?

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Looks like this needs some updates in order for the tests to pass and to get coverage for the `logp` implementation.

File: pymc3/distributions/multivariate.py (hunk 5 above)
Comment (zoj613) [reply pymc3/distributions/multivariate.py]: Yep, the sum of the values have to be `0` so I set the precision to 6 digits.

File: pymc3/distributions/multivariate.py [hunk 6]
```diff
@@ -2048,3 +2050,100 @@ def random(self, point=None, size=None):
...
```
Comment (ricardoV94) [inline pymc3/distributions/multivariate.py]:
```suggestion
        return bound(out, at.abs(at.sum(value)) < 1e-06)
```

**Comment by ricardoV94** [comment]
Closing this as stale.

@zoj613 let us know if you want to pick this up at another time. Probably pymc-experimental would be a better target

**Pull request closed without merging**